### PR TITLE
ec2_vol: boto3 migration (migrating ansible/ansible#45500 to amazon.aws)

### PR DESCRIPTION
**Pull request by falcon78921**
##### SUMMARY

In order to keep the boto3 migration work in ansible/ansible#45500 alive, I attempted to add
the changes that were reviewed previously into a new PR.

ansible/ansible#45500 and some of the changes made in this PR were courtesy of @zeenlym.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

plugins/modules/ec2_vol.py

### CODE REVIEW

File: plugins/modules/ec2_vol.py [hunk 1]
```diff
@@ -379,125 +385,199 @@ def attach_volume(module, ec2, volume, instance):
...
+            final_tags = boto3_tag_list_to_ansible_dict(response.get('Tags'))
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't describe tags")
+
+    return final_tags, changed
+
 
 def main():
-    argument_spec = dict(
+    argument_spec = ec2_argument_spec()
```
Comment (tremble) [inline plugins/modules/ec2_vol.py:573]: AnsibleAWSModule automatically adds the ec2_argument_spec arguments, this line isn't needed.

File: plugins/modules/ec2_vol.py [hunk 2]
```diff
@@ -379,125 +385,199 @@ def attach_volume(module, ec2, volume, instance):
...
+    argument_spec.update(dict(
         instance=dict(),
         id=dict(),
         name=dict(),
-        volume_size=dict(type='int'),
+        volume_size=dict(),
```
Comment (tremble) [inline plugins/modules/ec2_vol.py:578]: It's better to define the 'type' here id possible.

File: plugins/modules/ec2_vol.py (hunk 1 above)
Comment (falcon78921) [reply plugins/modules/ec2_vol.py]: Thanks for taking a peek @tremble :)

I'm new to `ec2_vol` and Ansible dev in general. I'm still learning how things are being referenced, so if I made any stupid mistakes, please let me know.

I noticed `ec2_argument_spec` returns whatever `aws_common_argument_spec` returns, just with an updated region. I'm thinking the goal here is to then update `ec2_argument_spec`'s dictionary with whatever is passed from module args. Is that right?

File: plugins/modules/ec2_vol.py (hunk 2 above)
Comment (falcon78921) [reply plugins/modules/ec2_vol.py]: I added `type='int'` back into the dictionary.

File: plugins/modules/ec2_vol.py (hunk 1 above)
Comment (falcon78921) [reply plugins/modules/ec2_vol.py]: @tremble ah ok, so we should be good just calling `ec2_argument_spec.update()`, which will update the dictionary that's returned and then pass it to `AnsibleAWSModule`.

Comment (tremble) [reply plugins/modules/ec2_vol.py]: Even easier than that.

```
argument_spec = dict (.....)
module = AnsibleAWSModule(argument_spec=argument_spec
```

Comment (tremble) [reply plugins/modules/ec2_vol.py]: The 'update' pieces happen automatically inside AnsibleAWSModule (which in turn wraps AnsibleModule)

Comment (falcon78921) [reply plugins/modules/ec2_vol.py]: gotcha :+1:

Comment (falcon78921) [reply plugins/modules/ec2_vol.py]: ok, i just pushed the revision up.

**Comment by falcon78921** [comment]
Just wanted to say I'm still working on this PR.

**Comment by falcon78921** [comment]
@tremble I've been working on integrating the paginator for `get_volume()` and `get_volumes()`, per the previous PR's review.

I'm able to gather results using something like this:

```
def getVolume():
    vols = []
    paginator = ec2_conn.get_paginator('describe_volumes')
    vols_response = paginator.paginate(Filters=[{'Name':'volume-id', 'Values':["{}".format(vol_id)]}])
    for p in vols_response:
        vols.append(p['Volumes'])
    return vols[0]
```

```
def getVolumes():
    vols = []
    paginator = ec2_conn.get_paginator('describe_volumes')
    vols_response = paginator.paginate(Filters=[{'Name':'attachment.instance-id', 'Values':["{}".format(instance)]}])
    for p in vols_response:
        vols.append(p['Volumes'])
    return vols
```

I'm just a little lost on how to integrate the preceding back into `ec2_vol.py`. The `Paginator` object returns an iterator, but I'm not sure if this may change behaviors in other functions.

**Comment by falcon78921** [comment]
@s-hertel Regarding https://github.com/ansible/ansible/pull/45500#discussion_r218858613, do you mean removing `changed = True` and just letting `ec2_conn.delete_volume()` run? If it fails, `botocore` can return why.

**Comment by falcon78921** [comment]
> Were there any other things you were specifically thinking needed done?

@tremble Nope, let me know if there's anything else you need to me to do.

**Comment by tremble** [comment]
Looking at the results from the shippable / integration tests it looks like something's not quite right with the 'changed' status.  See for example : https://app.shippable.com/github/ansible-collections/amazon.aws/runs/700/31/console

**Comment by falcon78921** [comment]
@tremble I ran the `ec2_vol` tests in `main/tests/integration/targets/ec2_vol/tasks/main.yml` against another EC2 environment. I was able to get the tests to pass when I:

- Removed the test for the `list` state, given that it looks deprecated.
- Changing the assertions for the `attachment_set` state from `attached` to `attaching`, seeing as the status was running into a race condition.
- Putting `ignore_errors: true` on the tasks that were trying to reference non-existent volumes. I'm thinking this could be handled better.

**Comment by tremble** [comment]
List state may be deprecated but it's got to stay for another 18 months or so, and adding bugs during the next 18months would still be considered a 'bug' for any official support processes.  So I'd rather we didn't remove the tests.

attached vs attaching I'd handle using `in ["attached", "attaching"]` rather than an `==` as it avoids the race condition.

With regards handling the missing volumes, you can catch the exception using `is_boto3_error_code('InvalidVolume.NotFound')` and handle as appropriate.

**Comment by falcon78921** [comment]
In my AWS environment, `ec2_vol` seems to work when running through tests in `main/tests/integration/targets/ec2_vol/tasks/main.yml`.

**Comment by tremble** [comment]
> In my AWS environment, ec2_vol seems to work when running through tests in main/tests/integration/targets/ec2_vol/tasks/main.yml.

Interesting, they're working for me.  I wonder if there's something going on due to the way CI works...

**Comment by falcon78921** [comment]
@tremble :thinking:

**Comment by tremble** [comment]
I've done some digging and it looks like the ec2 instance isn't reporting the new volume in its description.  This is after the waiter's done its thing.

By the time I run ec2_instance_info after the failed test it's showing up.  I **suspect** that this is a very subtle race condition in the AWS APIs.

**Comment by tremble** [comment]
So...  I've got something that seems to work: https://github.com/ansible-collections/amazon.aws/pull/178/commits/de6a699121b84818eeb0e9ec0cd6e938aca48881

Looks like this is a race condition of some variety:  Basically the waiter which looks at the **volume** returns with "I'm attached to something" before the **instance** description includes the new volume.

File: plugins/modules/ec2_vol.py [hunk 3]
```diff
@@ -223,153 +228,165 @@
...
 try:
-    import boto
-    import boto.ec2
-    import boto.exception
-    from boto.exception import BotoServerError
-    from boto.ec2.blockdevicemapping import BlockDeviceType
-    from boto.ec2.blockdevicemapping import BlockDeviceMapping
+    import botocore
 except ImportError:
     pass  # Taken care of by ec2.HAS_BOTO
```
Comment (tremble) [inline plugins/modules/ec2_vol.py:244]: Niggle: this is Handled by AnsibleAWSModule rather than ec2.HAS_BOTO

**Review by tremble — CHANGES_REQUESTED** [review]

File: plugins/modules/ec2_vol.py (hunk 3 above)
Comment (falcon78921) [reply plugins/modules/ec2_vol.py]: @tremble Added clarification on the imports.

**Comment by falcon78921** [comment]
@tremble Thanks for investigating the testing oddity. I applied your diff.

**Comment by tremble** [comment]
Thanks for working on getting this boto3 migration in :)

**Comment by falcon78921** [comment]
@tremble Looks like there's still one test failing, but the rest are passing! :tada:

File: changelogs/fragments/53-ec2_vol-boto3-port.yml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+    - ec2_vol - ported ec2_vol to use boto3
```
Comment (tremble) [inline changelogs/fragments/53-ec2_vol-boto3-port.yml:2]:
```suggestion
    - ec2_vol - ported ec2_vol to use boto3 (https://github.com/ansible-collections/amazon.aws/pull/53).
```

File: plugins/modules/ec2_vol.py [hunk 5]
```diff
@@ -225,151 +230,165 @@
...
-    return hasattr(boto, 'Version') and LooseVersion(boto.Version) >= LooseVersion('2.39.0')
+def delete_volume(module, ec2_conn, volume_id=None):
+    changed = False
+    if volume_id:
+        try:
+            ec2_conn.delete_volume(VolumeId=volume_id)
+            changed = True
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:  # pylint: disable=duplicate-except
+            if is_boto3_error_code('InvalidVolume.NotFound'):
+                module.exit_json(changed=False)
```
Comment (tremble) [inline plugins/modules/ec2_vol.py:336]:
```suggestion
        except is_boto3_error_code('InvalidVolume.NotFound'):
            module.exit_json(changed=False)
        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:  # pylint: disable=duplicate-except
```

**Review by tremble — CHANGES_REQUESTED** [review]
2 last minor changes and then I think we're ready to get this merged.

In theory we could also make more use of AWSRetry, but I'd rather see this merged and I'll CC you on the relevant change so you could see how it would work.

**Comment by falcon78921** [comment]
@tremble :+1:

**Review by tremble — APPROVED** [review]

**Comment by tremble** [comment]
Thanks for seeing this through.  Scratch one more module off the boto v2 list.

**Pull request merged**